modules/core/04-channel/keeper/upgrade.go

```diff
@@ -77,3 +81,112 @@ func (k Keeper) WriteUpgradeInitChannel(
 
 	emitChannelUpgradeInitEvent(ctx, portID, channelID, upgradeSequence, channelUpgrade)
 }
+
+func (k Keeper) ChanUpgradeTry(
+	ctx sdk.Context,
+	portID string,
+	channelID string,
+	chanCap *capabilitytypes.Capability,
+	counterPartyChannel types.Channel,
+	counterPartyUpgradeSequence uint64,
+	proposedUpgradeChannel types.Channel,
+	timeoutHeight clienttypes.Height,
+	timeoutTimestamp uint64,
+	channelProof []byte,
+	upgradeTimeoutProof []byte,
+	upgradeSequenceProof []byte,
+	proofHeight clienttypes.Height,
+) error {
+
+	channel, found := k.GetChannel(ctx, portID, channelID)
+	if !found {
+		return errorsmod.Wrapf(types.ErrChannelNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	if !collections.Contains(channel.State, []types.State{types.OPEN, types.INITUPGRADE}) {
+		return errorsmod.Wrapf(types.ErrInvalidChannelState, "expected one of [%s, %s], got %s", types.OPEN, types.INITUPGRADE, channel.State)
+	}
+
+	if !k.scopedKeeper.AuthenticateCapability(ctx, chanCap, host.ChannelCapabilityPath(portID, channelID)) {
+		return errorsmod.Wrapf(types.ErrChannelCapabilityNotFound, "caller does not own capability for channel, port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	if proposedUpgradeChannel.State != types.TRYUPGRADE || proposedUpgradeChannel.Counterparty.PortId != channel.Counterparty.PortId ||
+		proposedUpgradeChannel.Counterparty.ChannelId != channel.Counterparty.ChannelId {
+		return errorsmod.Wrap(types.ErrInvalidChannel, "proposed channel upgrade is invalid")
+	}
+
+	if strings.TrimSpace(proposedUpgradeChannel.Version) == "" {
+		return errorsmod.Wrap(types.ErrInvalidChannelVersion, "channel version must be not be empty")
+	}
+
+	if !channel.Ordering.SubsetOf(proposedUpgradeChannel.Ordering) {
+		return errorsmod.Wrap(types.ErrInvalidChannelOrdering, "channel ordering must be a subset of the new ordering")
+	}
+
+	if counterPartyChannel.Ordering != proposedUpgradeChannel.Ordering {
+		return errorsmod.Wrapf(types.ErrInvalidChannelOrdering, "channel ordering of counterparty channel and proposed channel must be equal")
+	}
+
+	connection, err := k.GetConnection(ctx, channel.ConnectionHops[0])
+	if err != nil {
+		return err
+	}
```

Comment:
Connection hops is an upgradable field, I think we should be operating on the `proposedUpgradeChannel` connection hops instead. If we are upgrading/altering the connection of the underlying channel end, then we should be proving counterparty state against the new connection. cc. @AdityaSripal 

```suggestion
	connection, err := k.GetConnection(ctx, proposedUpgradeChannel.ConnectionHops[0])
	if err != nil {
		return err
	}
```

Reply:
Addressed in https://github.com/cosmos/ibc-go/pull/3333/commits/165e23d0f275be6eafe3c416e5b659b7b4f05543